cmd/isp/src/protocol.rs

```diff
@@ -0,0 +1,544 @@
+    crc.digest(&response);
+
+    let digest = crc.get_crc();
+
+    if !(((digest & 0xff) == (frame.crc16_low as u16))
+        && (((digest >> 8) & 0xff) == (frame.crc16_high as u16)))
+    {
+        return Err(anyhow!(
+            "CRC failure on packet expect {:x}{:x} got {:x}",
+            frame.crc16_high as u8,
+            frame.crc16_low as u8,
+            digest
+        ));
+    }
+
+    Ok(())
+}
+
+pub fn read_data(port: &mut dyn serialport::SerialPort) -> Result<Vec<u8>> {
+    let mut frame_bytes = vec![0; core::mem::size_of::<FramingPacket>()];
+    let mut cnt = 0;
+
+    while cnt != core::mem::size_of::<FramingPacket>() {
+        let r = port.read(&mut frame_bytes[cnt..])?;
+        cnt += r;
+    }
+
+    let frame = FramingPacket::read_from(frame_bytes.as_slice())
+        .ok_or(anyhow!("parse_erorr"))?;
+
+    if frame.header.packet_type != (PacketType::Data as u8) {
+        return Err(anyhow!(
+            "Expected a data packet, got {:x} instead",
+            frame.header.packet_type
+        ));
+    }
+
+    cnt = 0;
+    let length: usize =
+        (frame.length_low as usize) | ((frame.length_high as usize) << 8);
+    let mut response = vec![0; length];
+
+    while cnt != length {
+        let r = port.read(&mut response[cnt..])?;
+        cnt += r;
+    }
+
+    check_crc(&frame_bytes, &response, &frame)?;
+
+    Ok(response)
+}
+
+pub fn read_response(
+    port: &mut dyn serialport::SerialPort,
+    response_type: ResponseCode,
+) -> Result<Vec<u32>> {
+    let mut frame_bytes = vec![0; core::mem::size_of::<FramingPacket>()];
+    let mut cnt = 0;
+
+    while cnt != core::mem::size_of::<FramingPacket>() {
+        let r = port.read(&mut frame_bytes[cnt..])?;
+        cnt += r;
+    }
+
+    let frame = FramingPacket::read_from(frame_bytes.as_slice())
+        .ok_or(anyhow!("Parse_err"))?;
+
+    // A response packet is a specific type of command packet.
+    if frame.header.packet_type != (PacketType::Command as u8) {
+        return Err(anyhow!(
+            "Expected a command, got {:x}",
+            frame.header.packet_type
+        ));
+    }
+
+    cnt = 0;
+    let length: usize =
+        (frame.length_low as usize) | ((frame.length_high as usize) << 8);
+    let mut response = vec![0; length];
+
+    while cnt != length {
+        let r = port.read(&mut response[cnt..])?;
+        cnt += r;
+    }
+
+    check_crc(&frame_bytes, &response, &frame)?;
+
+    let command =
+        RawCommand::read_from(&response[..core::mem::size_of::<RawCommand>()])
+            .ok_or(anyhow!("Parse error"))?;
+
+    if command.tag != (response_type as u8) {
+        return Err(anyhow!(
+            "Expected a response type of {:x}, got {:x}",
+            response_type as u8,
+            command.tag
+        ));
+    }
+
+    let mut params: Vec<u32> = Vec::new();
+    let mut cnt = 0;
+    let mut index = core::mem::size_of::<RawCommand>();
+
+    while cnt < command.parameter_count {
```

Comment:
Nit: You could use `params.len()` instead of `cnt`, to keep track of one fewer thing